File: 2024/day19/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

my $dir = dirname(__FILE__);
my $input_file = File::Spec->catfile($dir, '..', 'input.txt');

open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
my $input = do { local $/; <$fh> };
close $fh;

$input =~ s/\s+$//;

my ($patterns_str, $designs_str) = split /\n\n/, $input, 2;
my @patterns = map { s/^\s+|\s+$//gr } split /,/, $patterns_str;
my @designs = split /\n/, $designs_str;

sub can_form {
    my ($design) = @_;
    my %memo;

    my $dp;
    $dp = sub {
        my ($pos) = @_;
        return 1 if $pos == length($design);
        return $memo{$pos} if exists $memo{$pos};

        for my $pattern (@patterns) {
            my $plen = length($pattern);
            if (substr($design, $pos, $plen) eq $pattern) {
                if ($dp->($pos + $plen)) {
                    $memo{$pos} = 1;
                    return 1;
                }
            }
        }
        $memo{$pos} = 0;
        return 0;
    };

    return $dp->(0);
}

sub count_ways {
    my ($design) = @_;
    my %memo;

    my $dp;
    $dp = sub {
        my ($pos) = @_;
        return 1 if $pos == length($design);
        return $memo{$pos} if exists $memo{$pos};

        my $total = 0;
        for my $pattern (@patterns) {
            my $plen = length($pattern);
            if (substr($design, $pos, $plen) eq $pattern) {
                $total += $dp->($pos + $plen);
            }
        }
        $memo{$pos} = $total;
        return $total;
    };

    return $dp->(0);
}

sub part1 {
    my $count = 0;
    for my $design (@designs) {
        $count++ if can_form($design);
    }
    return $count;
}

sub part2 {
    my $total = 0;
    for my $design (@designs) {
        $total += count_ways($design);
    }
    return $total;
}

print "Part 1: ", part1(), "\n";
print "Part 2: ", part2(), "\n";

File: 2024/day19/go/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var patterns []string

func countWays(design string) int64 {
	n := len(design)
	memo := make(map[int]int64)

	var dp func(pos int) int64
	dp = func(pos int) int64 {
		if pos == n {
			return 1
		}
		if val, ok := memo[pos]; ok {
			return val
		}

		var total int64 = 0
		for _, pattern := range patterns {
			plen := len(pattern)
			if pos+plen <= n && design[pos:pos+plen] == pattern {
				total += dp(pos + plen)
			}
		}

		memo[pos] = total
		return total
	}

	return dp(0)
}

func part1(designs []string) int {
	count := 0
	for _, design := range designs {
		if countWays(design) > 0 {
			count++
		}
	}
	return count
}

func part2(designs []string) int64 {
	var total int64 = 0
	for _, design := range designs {
		total += countWays(design)
	}
	return total
}

func main() {
	exePath, _ := os.Executable()
	exeDir := filepath.Dir(exePath)
	inputPath := filepath.Join(exeDir, "..", "input.txt")

	// Fallback for running with go run
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		inputPath = filepath.Join(filepath.Dir(os.Args[0]), "..", "input.txt")
		if _, err := os.Stat(inputPath); os.IsNotExist(err) {
			// Try relative to current working directory
			inputPath = "../input.txt"
		}
	}

	data, err := os.ReadFile(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	content := strings.TrimSpace(string(data))
	parts := strings.Split(content, "\n\n")

	// Parse patterns
	patternParts := strings.Split(parts[0], ",")
	patterns = make([]string, len(patternParts))
	for i, p := range patternParts {
		patterns[i] = strings.TrimSpace(p)
	}

	// Parse designs
	designs := strings.Split(strings.TrimSpace(parts[1]), "\n")

	fmt.Printf("Part 1: %d\n", part1(designs))
	fmt.Printf("Part 2: %d\n", part2(designs))
}
